src/v/cluster/topic_table.cc

```diff
@@ -647,23 +647,40 @@ void topic_table::notify_waiters() {
     /// If by invocation of this method there are no waiters, notify
     /// function_ptrs stored in \ref notifications, without consuming all
     /// pending_deltas for when a subsequent waiter does arrive
+
     std::vector<delta> changes;
+
+    // \ref notify_waiters is called after every apply. Hence for the most
+    // part there should only be a few items towards the end of \ref
+    // _pending_deltas that need to be sent to callbacks in \ref
+    // notifications. \ref _last_consumed_by_notifier_offset allows us to
+    // skip ahead to the items added since the last \ref notify_waiters
+    // call.
+    auto starting_iter = _pending_deltas.begin()
+                         + _last_consumed_by_notifier_offset;
+
     std::copy_if(
-      _pending_deltas.begin(),
+      starting_iter,
       _pending_deltas.end(),
       std::back_inserter(changes),
       [this](const delta& d) { return d.offset > _last_consumed_by_notifier; });
-    for (auto& cb : _notifications) {
-        cb.second(changes);
-    }
+
     if (!changes.empty()) {
+        for (auto& cb : _notifications) {
+            cb.second(changes);
```

Comment:
by moving the for-loop under the changes.empty check i assume there are no consumers that are interested in the case that changes _is_ empty? i'd assume not, but it is a semantic change for the callback consumers.

Reply:
Right you are. Definitely a semantic change. I went through and checked all of the consumers of this callback and none do anything for an empty `_pending_deltas`. Should I maybe add a comment to make note of this behavior?

Reply:
the other semantic change is that late registration of a callback might result in missing some deltas notifications that were previously notified and now skipped.

these semantic changes are only important insomuch as the use cases are safe, so verifying that users are tolerant of the changes is important. we don't have to maintain the previous semantics since this is just internal api / utility.